### PR TITLE
 Added new target support to Boolean, String, Number Object

### PR DESCRIPTION
**Pull request by kisbg**
JerryScript-DCO-1.0-Signed-off-by: bence gabor kis kisbg@inf.u-szeged.hu

### CODE REVIEW

**Comment by kisbg** [comment]
@rerobika Thank you for the review!

**Comment by rerobika** [comment]
please rebase the PR to the latest master.

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 1]
```diff
@@ -554,12 +564,38 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
       JERRY_ASSERT (ecma_is_value_boolean (value));
-
-      return ecma_op_create_boolean_object (value);
+    #if ENABLED (JERRY_BUILTIN_BOOLEAN)
+      proto_id = ECMA_BUILTIN_ID_BOOLEAN_PROTOTYPE;
+    #endif /* !(ENABLED (JERRY_BUILTIN_BOOLEAN) */
+      lit_id = LIT_MAGIC_STRING_BOOLEAN_UL;
     }
   }
+  return ecma_op_create_class_object (proto_id, ecma_copy_value (value), lit_id);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:573]: Pass only the value here as second argument.

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 2]
```diff
@@ -538,12 +561,43 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
+/**
+ * ToObject operation.
+ *
+ * See also:
+ *          ECMA-262 v5, 9.9
+ *
+ * @return ecma value
+ *         Returned value must be freed with ecma_free_value
+ */
+ecma_value_t
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-conversion.c:585]: This request is still valid.

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 3]
```diff
@@ -538,12 +561,43 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
+                          ecma_value_t value, /**< ecma value */
+                          uint16_t lit_id) /**< magic string id */
+{
+  ecma_object_t *object_p = ecma_create_object (prototype_obj_p,
+                                                sizeof (ecma_extended_object_t),
+                                                ECMA_OBJECT_TYPE_CLASS);
+
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
+  ext_object_p->u.class_prop.class_id = lit_id;
+  ext_object_p->u.class_prop.u.value = value;
```
[the remark this answers is not in the document]
Comment (rerobika) [reply jerry-core/ecma/operations/ecma-conversion.c:596]: Ditto.

File: jerry-core/ecma/operations/ecma-boolean-object.c [hunk 4]
```diff
@@ -57,6 +73,13 @@ ecma_op_create_boolean_object (ecma_value_t arg) /**< argument passed to the Boo
   ext_object_p->u.class_prop.class_id = LIT_MAGIC_STRING_BOOLEAN_UL;
   ext_object_p->u.class_prop.u.value = ecma_make_boolean_value (boolean_value);
 
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_CONTEXT (current_new_target))
+  {
+    ecma_deref_object (prototype_obj_p);
+  }
+#endif /* !(ENABLED (JERRY_ESNEXT) */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-boolean-object.c:81]: `#endif /* (ENABLED (JERRY_ESNEXT) */`

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 5]
```diff
@@ -554,12 +564,39 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
     else
     {
       JERRY_ASSERT (ecma_is_value_boolean (value));
-
-      return ecma_op_create_boolean_object (value);
+    #if ENABLED (JERRY_BUILTIN_BOOLEAN)
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:567]: `#` conditions must be in the beginning of the line.

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 6]
```diff
@@ -554,12 +564,39 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
+      proto_id = ECMA_BUILTIN_ID_BOOLEAN_PROTOTYPE;
+    #endif /* !(ENABLED (JERRY_BUILTIN_BOOLEAN) */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:569]: `#endif /* ENABLED (JERRY_BUILTIN_BOOLEAN) */`

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 7]
```diff
@@ -554,12 +564,39 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
+                             ecma_value_t value, /**< ecma value */
+                             uint16_t lit_id) /**< magic string id */
+{
+  ecma_object_t *object_p = ecma_create_object (ecma_builtin_get (proto_id),
+                                                sizeof (ecma_extended_object_t),
+                                                ECMA_OBJECT_TYPE_CLASS);
+
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
+  ext_object_p->u.class_prop.class_id = lit_id;
+  ext_object_p->u.class_prop.u.value = ecma_copy_value (value);
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:597]: To be more universal, please use `ecma_copy_value_if_not_object` instead.

File: jerry-core/ecma/operations/ecma-conversion.c [hunk 8]
```diff
@@ -554,12 +564,39 @@ ecma_op_to_object (ecma_value_t value) /**< ecma value */
...
     }
   }
+
+  return ecma_op_create_class_object (proto_id, value, lit_id);
 } /* ecma_op_to_object */
 
+/**
+ * create class object operation.
+ *
+ * See also:
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-conversion.c:580]: No standard reference is needed.

File: jerry-core/ecma/operations/ecma-number-object.c [hunk 9]
```diff
@@ -63,7 +77,12 @@ ecma_op_create_number_object (ecma_value_t arg) /**< argument passed to the Numb
 
   /* Pass reference (no need to free conv_to_num_completion). */
   ext_object_p->u.class_prop.u.value = conv_to_num_completion;
-
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_CONTEXT (current_new_target))
+  {
+    ecma_deref_object (prototype_obj_p);
+  }
+#endif /* !(ENABLED (JERRY_ESNEXT) */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-number-object.c:85]: ditto.

File: jerry-core/ecma/operations/ecma-number-object.c [hunk 10]
```diff
@@ -42,18 +44,30 @@ ecma_value_t
...
+ #if ENABLED (JERRY_ESNEXT)
+  if (JERRY_CONTEXT (current_new_target))
+  {
+    prototype_obj_p = ecma_op_get_prototype_from_constructor (JERRY_CONTEXT (current_new_target), proto_id);
+    if (JERRY_UNLIKELY (prototype_obj_p == NULL))
+    {
+      return ECMA_VALUE_ERROR;
+    }
+  }
+ #endif /* !(ENABLED (JERRY_ESNEXT) */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-number-object.c:70]: ditto.

File: jerry-core/ecma/operations/ecma-string-object.c [hunk 11]
```diff
@@ -61,19 +65,34 @@ ecma_op_create_string_object (const ecma_value_t *arguments_list_p, /**< list of
...
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_CONTEXT (current_new_target))
+  {
+    prototype_obj_p = ecma_op_get_prototype_from_constructor (JERRY_CONTEXT (current_new_target), proto_id);
+    if (JERRY_UNLIKELY (prototype_obj_p == NULL))
+    {
+      return ECMA_VALUE_ERROR;
+    }
+  }
+#endif /* !(ENABLED (JERRY_ESNEXT) */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-string-object.c:82]: ditto.

File: jerry-core/ecma/operations/ecma-string-object.c [hunk 12]
```diff
@@ -61,19 +65,34 @@ ecma_op_create_string_object (const ecma_value_t *arguments_list_p, /**< list of
...
   ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
   ext_object_p->u.class_prop.class_id = LIT_MAGIC_STRING_STRING_UL;
   ext_object_p->u.class_prop.u.value = prim_value;
-
+#if ENABLED (JERRY_ESNEXT)
+  if (JERRY_CONTEXT (current_new_target))
+  {
+    ecma_deref_object (prototype_obj_p);
+  }
+#endif /* !(ENABLED (JERRY_ESNEXT) */
```
Comment (rerobika) [inline jerry-core/ecma/operations/ecma-string-object.c:95]: ditto.

**Pull request closed without merging**